### DISCUSSION TITLE
RFC: Next.js App Router HN demo

### DISCUSSION THREAD

**Discussion by trongthanh** [pmndrs/jotai#2552] [Ideas] [answered]
As more projects are adopting [Next.js 13+ App Router](https://nextjs.org/docs/app/building-your-application/routing) approach but current Next.js documentation and demo are still showcasing Page Router, I had derived the HN Demo to showcase how to use SSR, `useRehydrateAtoms` and Provider in a Next.js App Router app.

Here's [the demo on Stackblitz](https://stackblitz.com/edit/jotai-nextjs-app-router-demo?file=app%2Fpage.tsx,app%2Fabout%2Fpage.tsx,store%2Findex.ts,components%2FPost.tsx) (Demo A)

I intend to create a PR to either embed or link this demo to [Jotai's Next.js doc page](https://jotai.org/docs/guides/nextjs) to provide some reference for new learners. My only concern is the use of `store` API in the read atom where I need to store the postData to local cache. 

I also tried another approach to achieve [memory cache with write atom only](https://stackblitz.com/edit/stackblitz-starters-8dkavn?file=app%2Fpage.tsx) (Demo B). And feel that this one follow Jotai's wisdom better, but loosing the benefit of Suspense. 

Please review and suggest which is better as an official documentation demo.

**Comment by dai-shi** [comment]
Thanks for opening new discussion! App Router HN demo would be great.

Some random comments:
- Provider should be close to root. `layout.tsx` should be appropriate?
- We shouldn't depend on `store` API for the basic use case.
- The benefit of Suspense is important.

Are they not supported in the Page Router example?

**Reply by trongthanh** [reply]
> Provider should be close to root. layout.tsx should be appropriate?

Let me see to it. But from a few early trials, Provider must be inside a `use client` component. If not, I cannot provide own store via its attribute. Besides, another observation is that Jotai's hooks will not work in server component.

> We shouldn't depend on store API for the basic use case. The benefit of Suspense is important.

Agree. But in this specific case, I did revise the [original demo previously on Codesandbox](https://codesandbox.io/p/devbox/nextjs-with-jotai-async-forked-rgdddf) to make the cache works for client side as well. (When you click Previous, the post show immediately thanks to postCache atom). And to make the Suspense loading effect work, the `await fetch...` must be in an `async` read function. However, inside a read function, there's no `set` function to store the post cache. Therefore, the use of `store.set`.

> Are they not supported in the Page Router example?

The Page Router example works fine and is predictable. The behaviors of Jotai's hooks and utils in App Router's client components are pretty similar to Page Router's pages.

**Reply by dai-shi** [reply]
> Provider must be inside a use client component. 

Provider has `use client` directive in the library, so you don't need it on your end.

> I cannot provide own store via its attribute.

But, that's true if you need your own store.

> (When you click Previous, the post show immediately thanks to postCache atom).

That sounds like you are solving a new problem that is not strictly related with Next.js.
We have an experimental library `jotai-cache` to address the issue.
So, I'd suggest you don't solve the cache problem with the Next.js example.

**Reply by trongthanh** [reply]
> Provider has use client directive in the library, so you don't need it on your end.

This is a new realization for me. Then, I indeed shouldn't place it close to root, because the Provider will make all of its children client components. In an App Router app, there should be a boundary between server & client components. The whole purpose of App Router is to leveraging RSC and reducing client-side bundles. If Jotai is a client component ONLY, then Provider should start from the client components root. It's worth noting that client components are actually still SSR'ed and Jotai is still working at server side just like page router.

Lemme put the `app/page.tsx` code here to explain more:

```tsx
export default async function IndexPage() {
  const initialPost = await fetchPost();

  return (
    <>
      <nav>
        <Link href="/about">About</Link>
      </nav>
      <h1>Index</h1>
      <article>
        <Provider>
          <Post initialPost={initialPost} />
        </Provider>
      </article>
    </>
  );
}
```

So, this page is a server component, and `<Provider>` marks the boundary of client component root. This page will be SSR without a JS bundle. The page template will be rehydrated at client-side with [RSC Payload](https://nextjs.org/docs/app/building-your-application/rendering/server-components#how-are-server-components-rendered). The `<Provider>` and `<Post>` will also be SSR with prop's contents serialized into RSC Payload. This is why we cannot pass a custom store to Provider prop here. The prop value passed down to client components must be serializable. Next.js will bundle any JS needed for the client components and will use the JS bundle, together with the prop data from server component to rehydrate the client component tree.

In short, IMO this is a good example for the Next.js App Router use case which shows where Jotai off-limits.

Now back to the cache problem, the reason I tried to fix the 'Prev' link cache is because in the original demo, postCache atom is used only once for hydration so it's a bit confusing. Also, there was no 'Prev' link. (I added it to the new demo). So if we need to avoid the direct store API, I'll revise this demo simpler with better atom naming.

**Reply by dai-shi** [reply]
> because the Provider will make all of its children client components.

No, `children` can be server components.

**Reply by trongthanh** [reply]
> > because the Provider will make all of its children client components.

> No, children can be server components.

Unfortunately, it's not. Once a parent `use client`, all of its imports and children are part of the client bundle. Here's the related excerpt from [Next.js doc](https://nextjs.org/docs/app/building-your-application/rendering/client-components):

<img width="665" alt="Screenshot 2024-05-16 at 10 25 21 PM" src="https://github.com/pmndrs/jotai/assets/234226/18af1d0e-c9a2-4c0e-bc8a-33aed322371e">

**Reply by dai-shi** [reply]
Pass a server component to Provider's children.

```jsx
// server component

export function Layout() {
  return <Provider><InnerServerComponent /></Provider>
}

// Or just children
export function Layout({ children }) {
  return <Provider>{children}</Provider>
}
```

Don't add `"use client"` in this code.

**Reply by trongthanh** [reply]
I see.

I have moved the `<Provider>` to layout.tsx and revise the `postData` read to avoid `store` API. `postCache` is only used for hydration data.

Please check if [the demo](https://stackblitz.com/edit/jotai-nextjs-app-router-demo?file=app%2Fpage.tsx,app%2Fabout%2Fpage.tsx,store%2Findex.ts,components%2FPost.tsx) is OK now.

**Reply by dai-shi** [reply]
Yeah, it looks good to me!

**Reply by thanhtran-ct** [reply]
Thanks. I'll create the PR, intended to place it in the Next.js guide. Should this be an embed demo or a link? There's already 2 embed demos in that page.

PS. this is my other profile, mistaken when replied.

**Reply by dai-shi** [reply, marked as answer]
Good point. I think it would be nice to have one embed demo and rest are links.
